Foundation/Network/HostName.hsc

```diff
@@ -0,0 +1,166 @@
+-- |
+-- Module      : Foundation.Network.HostName
+-- License     : BSD-style
+-- Maintainer  : Nicolas Di Prima <nicolas@primetype.co.uk>
+-- Stability   : experimental
+-- Portability : portable
+--
+-- HostName and HostName info
+--
+-- > getHostNameInfo "github.com" :: IO (HostNameInfo IPv4)
+--
+-- > getHostNameInfo "google.com" :: IO (HostNameInfo IPv6)
+--
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Foundation.Network.HostName
+    ( HostName(..)
+    , HostNameInfo(..)
+    , getHostNameInfo
+    , getHostNameInfo_
+    ) where
+
+import Foundation.Class.Storable
+import Foundation.Internal.Base
+import Foundation.Internal.Proxy
+import Foundation.Hashing (Hashable)
+import Foundation.String
+import Foundation.Network.Address
+
+import Foreign.C.String
+import Foreign.C.Types
+import Foreign.Ptr (nullPtr)
+import Control.Concurrent.MVar
+import System.IO.Unsafe (unsafePerformIO)
+
+#ifdef mingw32_HOST_OS
+#include <winsock2.h>
+#else
+#include "netinet/in.h"
+#include "netdb.h"
+#endif
+
+-- | HostName
+--
+newtype HostName = HostName { toString :: String }
+  deriving (Show, Eq, Ord, Typeable, Hashable)
```

Comment:
I think it make sense to not derive Show automatically, as this is a primitive type. hostname should be displayed as if it was a String itself